### PR TITLE
Replace LoadOrStore with CmpAndSwap

### PR DESCRIPTION
**Pull request by dopey**

### CODE REVIEW

File: bolt/bbolt.go [hunk 1]
```diff
@@ -118,32 +118,49 @@ func (db *DB) List(bucket []byte) ([]*database.Entry, error) {
...
+			return nil, false, errors.Wrapf(err, "failed to rollback read-only CmpAndSwap transaction on %s/%s", bucket, key)
+		}
+		return val, swapped, err
+	}
+}
+
+func cmpAndSwap(boltBucket *bolt.Bucket, key, newValue, oldValue []byte) ([]byte, bool, error) {
+	current := boltBucket.Get(key)
+	if !bytes.Equal(current, oldValue) {
+		return current, false, nil
```
Comment (maraino) [inline bolt/bbolt.go]: You might need to return a copy of `current`. According to bbolt docs:
```
func (b *Bucket) Get(key []byte) []byte

Get retrieves the value for a key in the bucket. Returns a nil value if the key
does not exist or if the key is a nested bucket. The returned value is only
valid for the life of the transaction.
```
Check if this is properly done in the other implementations

File: badger/badger.go [hunk 2]
```diff
@@ -211,33 +211,50 @@ func (db *DB) List(bucket []byte) ([]*database.Entry, error) {
...
+		if oldValue != nil {
+			return nil, false, nil
+		}
+		// For any error other than 'NotFound' return the error w/out swapping.
+	} else if err != nil {
+		return nil, false, err
+		// If expected and current value don't match then return w/out swapping values.
+	} else if !bytes.Equal(current, oldValue) {
+		return current, false, nil
+	}
```
Comment (maraino) [inline badger/badger.go]: Ok, but you can simplify it as you did in bolt if badgerGet returns nil:
```go
if err != nil && !database.IsErrNotFound(err) {
        return nil, false, err
}
if !bytes.Equal(current, oldValue){
       return current, false, nil
}
```

File: database/database.go [hunk 3]
```diff
@@ -86,10 +86,10 @@ type DB interface {
...
 	// Set sets the given value in the given table/bucket and key.
 	Set(bucket, key, value []byte) error
-	// LoadOrStore returns the value stored in the given table/bucket and key
-	// if it exists, otherwise it sets the value. Returns 'true' if the value
-	// was found, false otherwise.
-	LoadOrStore(bucket, key, value []byte) ([]byte, bool, error)
+	// CmpAndSwap swaps the value at the given bucket and key if the current
+	// value is equivalent to the oldValue input. Returns 'true' if the
+	// swap was successful and 'false' otherwise.
+	CmpAndSwap(bucket, key, value, oldValue []byte) ([]byte, bool, error)
```
Comment (maraino) [inline database/database.go]: Swap new and old.

File: mysql/mysql.go [hunk 4]
```diff
@@ -128,40 +131,52 @@ func (db *DB) List(bucket []byte) ([]*database.Entry, error) {
...
 		}
-		return nil, false, nil
-	case err != nil:
+	} else if err != nil {
 		return nil, false, errors.Wrapf(err, "failed to get %s/%s", bucket, key)
-	default:
-		return res, true, nil
+	} else if !bytes.Equal(current, oldValue) {
+		return current, false, nil
 	}
```
Comment (maraino) [inline mysql/mysql.go]: You can simplify this if/else/else if as before

File: bolt/bbolt.go (hunk 1 above)
Comment (dopey) [reply bolt/bbolt.go]: good catch.

**Comment by dopey** [comment]
Think I made necessary changes for all the comments.

**Pull request merged**